exercises/nondet_monad_cont/descr.md

```diff
@@ -0,0 +1,522 @@
+`interleave` is defined in such a way that `interleave evens odd`
+produces the infinite sequence `0, 1, 2, 3, 4, 5, ...`.
+It is *fair* in the sense that each branch in turn is allowed
+to produce a result.
+
+The constructor function `(>>-)` has the same type as `(>>=)`.
+It is a *fair sequencing* operator.
+Indeed, a problem with ordinary sequencing `(>>=)` is that
+it gives rise to ordinary (unfair) disjunctions.
+To see this, suppose that the left-hand argument of `(>>=)`
+is a computation that produces `x` as its first result,
+followed with a computation `m1` that may produce more results.
+Thus, this left-hand argument is equivalent to
+`choose (return x) m1`.
+When we sequentially compose it with a computation `m2`,
+we obtain
+`(choose (return x) m1) >>= m2`,
+which is equivalent to
+`choose (return x >>= m2) (m1 >>= m2)`,
+which itself is the same as
+`choose (m2 x) (m1 >>= m2)`.
+We are faced with an ordinary (unfair) disjunction.
+The problem, again, is that if `m2 x` produces an
+infinite number of results, then `m1 >>= m2`
+is never executed.
+To remedy this problem,
+the fair sequencing operator `(>>-)`
+is defined in such a way that
+`choose (return x) m1 >>- m2`
+is equivalent to
+`interleave (m2 x) (m1 >>- m2)`.
+Thus, it gives rise to a fair disjunction.
+
+## The `Seq` API
+
+The type of **on-demand sequences** is defined in a module named `Seq`.
+Beginning with version 4.07,
+this module is part of OCaml's standard library.
+
+```
+module Seq : sig
+
+  type 'a t = unit -> 'a node
+
+  and +'a node =
+  | Nil
+  | Cons of 'a * 'a t
+
+  val nil : 'a t
+  val cons: 'a -> 'a t -> 'a t
+  val singleton: 'a -> 'a t
+
+  val map: ('a -> 'b) -> 'a t -> 'b t
+  val concat: 'a t -> 'a t -> 'a t
+  val flatten: 'a t t -> 'a t
+
+  val take: int -> 'a t -> 'a t
+
+  val head: 'a t -> 'a option
+
+  val of_list: 'a list -> 'a t
+  val to_list: 'a t -> 'a list
+
+end
+```
+
+This data type is closely related to the algebraic data type of lists.
+Indeed, if instead of `unit -> 'a node` one had written just `'a node`,
+then this data type would have been isomorphic to the type of lists.
+
+The presence of `unit -> ...` indicates that a sequence is in fact a function.
+Calling this function, by applying it to the value `()`, amounts to requesting
+the head of the sequence. This head can be either `Nil`, which means that the
+sequence is empty, or `Cons (x, xs)`, which means that the first element of
+the sequence is `x` and the remaining elements form another sequence `xs`. It
+is worth noting that `xs` is itself a function, so the elements of the
+sequence `xs` need not be explicitly computed until `xs` is applied.
+
+Sequences are closely related to *iterators* in object-oriented languages,
+such as C++ and Java. Yet, sequences are much simpler than iterators, for
+two reasons:
+
+* they involve no mutable state;
+
+* they are just as easy to construct and to use as ordinary lists.
```

Comment:
Notice though that sequences cannot be observed as easily with pattern-matching since they are not compatible with nested patterns.

Reply:
True...